NitroxClient/MonoBehaviours/Gui/Chat/PlayerChat.cs

```diff
@@ -0,0 +1,144 @@
+using UnityEngine;
+using UnityEngine.EventSystems;
+using UnityEngine.UI;
+
+namespace NitroxClient.MonoBehaviours.Gui.Chat
+{
+    public class PlayerChat : uGUI_InputGroup
+    {
+        private const int LINE_CHAR_LIMIT = 255;
+        private const int MESSAGES_LIMIT = 64;
+        private const float TOGGLED_TRANSPARENCY = 0.4f;
+        public const float CHAT_VISIBILITY_TIME_LENGTH = 10f;
+
+        public static bool IsLoading { get; private set; } = true;
+
+        private PlayerChatManager playerChatManager;
+        private CanvasGroup canvasGroup;
+        private HorizontalOrVerticalLayoutGroup[] layoutGroups;
+        private GameObject logEntryPrefab;
+        private Image[] backgroundImages;
+        private bool transparent;
+        private InputField inputField;
+        public string inputText
+        {
+            get { return inputField.text; }
+            set { inputField.text = value; }
+        }
+
+        private static readonly Queue<ChatLogEntry> entries = new Queue<ChatLogEntry>();
+
+        public IEnumerator SetupChatComponents()
+        {
+            playerChatManager = NitroxServiceLocator.LocateService<PlayerChatManager>();
+
+            canvasGroup = GetComponent<CanvasGroup>();
+            layoutGroups = GetComponentsInChildren<HorizontalOrVerticalLayoutGroup>();
+
+            logEntryPrefab = GameObject.Find("ChatLogEntryPrefab");
+            logEntryPrefab.AddComponent<PlayerChatLogItem>();
+            logEntryPrefab.SetActive(false);
+
+            GetComponentsInChildren<Button>()[0].onClick.AddListener(ToggleBackgroundTransparency);
+            GetComponentsInChildren<Button>()[1].gameObject.AddComponent<PlayerChatPinButton>();
+
+            inputField = GetComponentInChildren<InputField>();
+            inputField.gameObject.AddComponent<PlayerChatInputField>().InputField = inputField;
+            inputField.GetComponentInChildren<Button>().onClick.AddListener(playerChatManager.SendMessage);
+
+            backgroundImages = new[]
+            {
+                transform.GetChild(0).GetComponent<Image>(), transform.GetChild(1).GetComponent<Image>(), transform.GetChild(3).GetComponent<Image>()
+            };
+
+            yield return new WaitForEndOfFrame(); //Needed so Select() works on initialization
+            IsLoading = false;
+        }
+
+        public void WriteLogEntry(string playerName, string message, Color color)
+        {
+            if (entries.Count == MESSAGES_LIMIT)
+            {
+                Destroy(entries.Dequeue().EntryObject);
+            }
+
+            ChatLogEntry chatLogEntry;
+            GameObject chatLogEntryObject;
+            if (entries.Count != 0 && entries.Last().PlayerName == playerName)
+            {
+                chatLogEntry = entries.Last();
+                chatLogEntry.MessageText += "\n" + message;
+                chatLogEntry.UpdateTime();
+                chatLogEntryObject = chatLogEntry.EntryObject;
```

Comment:
Would it be better to make each chat entry immutable instead of appending to the players last message?   Might simplify the logic here and be more intuitive.

Reply:
Yeah I can do this with two different prefabs (one for messages with names and one without).

Reply:
Not appending to the last message means that Unity needs to do more work handling the UI objects. I think it's worth it to append.